twisted/twisted
# [10109] Add nodeps and  fix no-ipv6 job.

## Scope and purpose

This adds the nodeps to the CI matrix.

I tried to refactor the GHA job to reuse as much as possible from the code...

The result is a bit harder to read...so I am ok to go back to code duplication in test.yaml file.

The `nodeps` run ironical, still has `coverage` as a dependency so that we can report the coverage and make sure we cover the extra code branches.

------

I am trying to reduce the size of the matrix so I went with this:

* latest 3.7 is used for non-ipv6 tests
* latest 3.6 used for nondeps tests
* minumim 3.6 and the rest are used for with ipv6 and with all deps tests

------

As a drive by, it fixes the no IPV6 tests as there were not working...
They were not working at lease since they were migrated to GHA - https://github.com/twisted/twisted/pull/1407/files
I have reviewed that PR but I have no idea why I have approved it :( ... I am trying to fix that damage here.

See coverage report which shows that ipv6 skip was never reached

https://codecov.io/gh/twisted/twisted/src/trunk/src/twisted/internet/test/test_tcp.py#L111

I have added a better description to ipv6 skip reason, as it was not that easy to understand what was going on there.

-------

As a drive by, I have used shorter names for GHA job names as with  older names it was harder to see what is failing

![image](https://user-images.githubusercontent.com/204609/109563276-75551680-7ad7-11eb-9813-2bfb07d4e22a.png)


## How to test

For no-IPV6 job.
Check that loopback ipv6 is down 
https://github.com/twisted/twisted/pull/1533/checks?check_run_id=2008044713#step:6:14

Then check the logs and see that tests are skipped due to

```
Does not work on systems without IPv6 support.
and
IPv6 not available. [Errno 99] Cannot assign requested address
```


At least the diff from this PR has coverage  https://app.codecov.io/gh/twisted/twisted/compare/1533/diff
so it looks like noipv6 code is triggerred.

------

For nodeps check... check the logs and you should see a lot of skipped test.
In theory, the coverage report should see that non SSL code branches are also covered now

For trunk https://codecov.io/gh/twisted/twisted/src/trunk/src/twisted/internet/tcp.py#L42

And check the similar link for this branch... 

https://app.codecov.io/gh/twisted/twisted/compare/1533/tree/src/twisted/internet/tcp.py#D1L43

--------

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10109
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
